mlir/docs/Rationale/RationaleLinalgDialect.md

```diff
@@ -506,6 +506,65 @@ potential by introducing lower-level IR ops and *smaller* Linalg ops.
 This gradually reduces the potential, all the way to Loops + VectorOps
 and LLVMIR.
 
+### Interchangeability of Forms<a name="forms"></a>
+
+Linalg's various forms (named, generic) also carry information, and that
+information should be preserved as much as possible during the progressive
+lowering. A `matmul` operation is a special case of a `contract` operation,
+which in turn is a special case of `generic` operation. Transformations on
+the more special forms should not be converted to the more generic ones
+unnecessarily, in the same way that they should not be broken down into
+loops + arithmetic if they can still be represented as a Linalg op.
```

Comment:
Shouldn't this paragraph come _after_ "Generic", "Category", "Named" are defined?

Reply:
OK, makes sense. But then, I wouldn't make any references to e.g. "named" or "generic" until these forms are defined. Otherwise it's confusing.

Instead, IMHO, you want to define what "forms" are, e.g. 
> Semantically equivalent representations of Linalg Ops,  i.e. two Linalg Ops OP_1 and OP_2 that given identical input, generate identical output 

(or something like this). Once that's defined, it should be sufficient to discuss the rationale.

Reply:
Reordered as requested.